d2/src/main/java/com/linkedin/d2/xds/balancer/XdsDirectory.java

```diff
@@ -0,0 +1,150 @@
+package com.linkedin.d2.xds.balancer;
+
+import com.linkedin.common.callback.Callback;
+import com.linkedin.d2.balancer.Directory;
+import com.linkedin.d2.xds.XdsClient;
+import indis.XdsD2;
+import io.grpc.Status;
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.List;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicReference;
+
+import static com.linkedin.d2.xds.XdsClient.*;
+
+
+public class XdsDirectory implements Directory
+{
+  private final XdsClient _xdsClient;
+  private final ConcurrentMap<String, String> _serviceNames = new ConcurrentHashMap<>();
+  private final ConcurrentMap<String, String> _clusterNames = new ConcurrentHashMap<>();
```

Comment:
Since this is just being used as a set, you can do this:
```suggestion
  private final Set<String> _serviceNames = ConcurrentHashMap.newKeySet();
  private final Set<String> _clusterNames = ConcurrentHashMap.newKeySet();
```

Reply:
Actually, if you go with the `ReadWriteLock` approach, these don't need even need to be concurrent

Reply:
Nevermind, they still do because `onChanged` doesn't always grab the lock

Reply:
"Since this is just being used as a set" --- they're definitely used as a map when being updated. We need to keep track of which name is mapped for which d2 service/cluster, so that when the name changes for that d2 service/cluster, the old value can also be removed while new value is added (when calling `_serviceNames.put(resourceName, nameData.getServiceName());`). If we use a set, the old value will be left in the set.